Rules/UseUsingScopeModifierInNewRunspaces.cs

```diff
@@ -0,0 +1,430 @@
+        /// prefixed with the 'Using' scope modifier. If so, analyze the block and generate diagnostic records for variables where it is missing.
+        /// </summary>
+        /// <param name="scriptBlockExpressionAst"></param>
+        /// <returns>
+        /// AstVisitAction.Continue or AstVisitAction.SkipChildren, depending on what we found. Diagnostic records are saved in `_diagnosticAccumulator`.
+        /// </returns>
+        public override AstVisitAction VisitScriptBlockExpression(ScriptBlockExpressionAst scriptBlockExpressionAst)
+        {
+            if (!(scriptBlockExpressionAst.Parent is CommandAst commandAst))
+            {
+                return AstVisitAction.Continue;
+            }
+
+            var cmdName = commandAst.GetCommandName();
+            if (cmdName == null)
+            {
+                // Skip for situations where command name cannot be resolved like `& $commandName -ComputerName -ScriptBlock { $foo }` 
+                return AstVisitAction.Continue;
+            }
+            
+            var scriptBlockParameterAst = commandAst.CommandElements[
+                    commandAst.CommandElements.IndexOf(scriptBlockExpressionAst) - 1] as CommandParameterAst;
+
+            if (IsInlineScriptBlock(cmdName) || 
+                IsJobScriptBlock(cmdName, scriptBlockParameterAst) ||
+                IsForeachScriptBlock(cmdName, scriptBlockParameterAst) ||
+                IsInvokeCommandComputerScriptBlock(cmdName, commandAst) ||
+                IsDSCScriptResource(cmdName, commandAst))
+            {
+                AnalyzeScriptBlock(scriptBlockExpressionAst);
+                return AstVisitAction.SkipChildren;
+            }
+
+            if (IsInvokeCommandSessionScriptBlock(cmdName, commandAst))
+            {
+                var sessionName = GetSessionName(commandAst);
+
+                var varsInLocalAssignments = FindVarsInAssignmentAsts(scriptBlockExpressionAst);
+                if (varsInLocalAssignments != null)
+                {
+                    AddAssignedVarsToSession(sessionName, varsInLocalAssignments);
+                }
+
+                GenerateDiagnosticRecords(
+                    FindNonAssignedNonUsingVarAsts(
+                        scriptBlockExpressionAst, 
+                        GetAssignedVarsInSession(sessionName)));
+
+                return AstVisitAction.SkipChildren;
+            }
+            
+            return AstVisitAction.Continue;
+        }
+
+        /// <summary>
+        /// FindVarsInAssignmentAsts: Retrieves all assigned variables from an Ast:
+        /// Example: `$foo = "foo"` ==> the VariableExpressionAst for $foo is returned
+        /// </summary>
+        /// <param name="ast"></param>
+        private static IEnumerable<VariableExpressionAst> FindVarsInAssignmentAsts(Ast ast)
+        {
+            // Find all variables that are assigned within this ast
+            return ast.FindAll(

```

Comment:
Converting away from this is really what I meant by converting to an AstVisitor; implementing methods like `VisitAssignmentStatement`.

But I won't block on it.

Instead, I think you should do the following:

- Convert the predicate to a static method
- Look for assignment ASTs rather than variable ASTs
- Use a `foreach` loop rather than LINQ

Reply:
I'm going to look into it for both `FindVarsInAssignmentAsts` and `FindNonAssignedNonUsingVarAsts`. I'm having a hard time wrapping my head around this, but will try.

Reply:
It's stateful and easier to get wrong. I think keeping the `FindAll` method is fine, you just want to turn your predicates into static methods, so they don't have to allocate closures

Reply:
I have given this some more thought. I feel if I use a visitor for this, I have to keep track of scope for each of the variables I encounter, which will become complicated and error prone real quick. I think I'll go the easier route then, and keep the `FindAll` but swap out the predicate for a static method as you explained (and invert the search to assignments).

Reply:
Yeah I agree

Reply:
I think I got this one, see latest commit.